### PR TITLE
Task-9: Add parameters to the query

### PR DESCRIPTION
**Pull request by Astoneti**
- Added parameters to the query.
- Refactored code.

### CODE REVIEW

File: delivery/meteorRoutes.js [hunk 1]
```diff
@@ -1,15 +1,19 @@
...
-    const data = await getMeteorData();
-    res.json(data);
+    const { date, count, wereDangerousMeteors } = req.query
+
+    const meteorData = await fetchMeteor(date, count, wereDangerousMeteors);
+  
+    res.json(meteorData)
   } catch (err) {
-    next(err);
+    next(new Exception(err.code || 500, err.message || 'An error occurred'))
```
Comment (Mikhail1992) [inline delivery/meteorRoutes.js:15]: previous variant was correct :)

File: delivery/meteorRoutes.js [hunk 2]
```diff
@@ -1,15 +1,19 @@
...
-const router = express.Router();
+const meteorRouter = express.Router()
 
-router.get('/', async (req, res, next) => {
+meteorRouter.get('/', async (req, res, next) => {
   try {
-    const data = await getMeteorData();
-    res.json(data);
+    const { date, count, wereDangerousMeteors } = req.query
+
```
Comment (Mikhail1992) [inline delivery/meteorRoutes.js:10]: you need to transform types right here, before passing parameters for fetchMeteor  service

File: helpers/meteorHelper.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    const isWereDangerousMeteors = wereDangerousMeteors; // No need to convert again
+    meteors = meteors.filter(meteor => 
+      meteor.is_potentially_hazardous_asteroid === isWereDangerousMeteors
+    );
+  }
+
+  if (count != null && !isNaN(count) && count > 0) {
+    meteors = meteors.slice(0, count); // Limit the number of meteors returned
+  }
+
```
Comment (Mikhail1992) [inline helpers/meteorHelper.js:50]: you don't need to filter meteors, you need to expand the response if you receive query parameter

File: helpers/meteorHelper.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  return meteors;
+}
+
+
+export function getStartAndEndDates(date) {
+  const givenDate = date ? new Date(date) : new Date()
```
Comment (Mikhail1992) [inline helpers/meteorHelper.js:56]: date - can be array with 2 items [startDate, endDate]

File: repository/meteorRepository.js [hunk 5]
```diff
@@ -1,24 +1,26 @@
...
   try {
-    const response = await axios.get(url, { httpsAgent: agent });
+    const response = await axios.get(url, { httpsAgent: agent })
     return response.data.near_earth_objects;
   } catch (err) {
-    console.error('Error fetching data from NASA API:', err.message);
-    throw new Exception(500, 'Failed to fetch data from NASA API');
+    console.error('Error fetching data from NASA API:', err.message)
+    throw new Exception(500, 'Failed to fetch data from NASA API')
   }
```
Comment (Mikhail1992) [inline repository/meteorRepository.js:21]: we don't need try..catch here

Comment (Mikhail1992) [reply repository/meteorRepository.js:21]: try..catch should  be added only on controller layer

File: useCases/fetchMeteorData.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  const dateFilter = new Date(date);
+  const filteredByDate = mappedMeteors.filter(meteor => {
+    const meteorDate = new Date(meteor.close_approach_date_full)
+    return meteorDate.toISOString().split('T')[0] === dateFilter.toISOString().split('T')[0]
+  })
+
+  const filteredMeteors = filterMeteors(filteredByDate, isCount, isWereDangerousMeteors)
+
+  if (isCount) {
+    return { count: filteredMeteors.length > isCount ? isCount : filteredMeteors.length }
```
Comment (Mikhail1992) [inline useCases/fetchMeteorData.js:21]: you need to extend initial response here

File: useCases/fetchMeteorData.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  if (isWereDangerousMeteors) {
+    const hasDangerous = filteredMeteors.some(meteor => meteor.is_potentially_hazardous_asteroid)
+    return { wereDangerousMeteors: hasDangerous }
```
Comment (Mikhail1992) [inline useCases/fetchMeteorData.js:26]: similar one

File: app/helpers/meteorHelper.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
+export function mapMeteorData(data) {
+  return Object.values(data).flatMap((meteors) =>
+    meteors.map((item) => {
+      const { relative_velocity, close_approach_date_full } =
+        extractCloseApproachData(item.close_approach_data)
```
Comment (EugeniBykovski) [inline app/helpers/meteorHelper.js:5]: Me can check on empty array also here:
(item.close_approach_data || []);

File: app/helpers/meteorHelper.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+      return {
+        id: item.id,
+        name: item.name,
+        diameter: getDiameter(item.estimated_diameter),
```
Comment (EugeniBykovski) [inline app/helpers/meteorHelper.js:10]: and here like empty object:
(item.estimated_diameter || {}),

File: app/helpers/meteorHelper.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        is_potentially_hazardous_asteroid:
+          item.is_potentially_hazardous_asteroid,
```
Comment (EugeniBykovski) [inline app/helpers/meteorHelper.js:12]: and here like ensure boolean: !!item.is_potentially_hazardous_asteroid

File: app/helpers/meteorHelper.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+  return {
+    relative_velocity:
+      closeApproachData[0].relative_velocity.kilometers_per_second,
+    close_approach_date_full: closeApproachData[0].close_approach_date_full,
+  }
+}
+
+const getDiameter = (diameter) => {
+  return (
```
Comment (EugeniBykovski) [inline app/helpers/meteorHelper.js:33]: also we can implement here small refactoring:
` if (!diameter?.meters) return null;
  const { estimated_diameter_min, estimated_diameter_max } = diameter.meters;
  return (estimated_diameter_min + estimated_diameter_max) / 2 || null;`

File: app/helpers/meteorHelper.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+export function filterMeteors(meteors, count, wereDangerousMeteors) {
+  if (wereDangerousMeteors != null) {
+    const isDangerousMeteors = wereDangerousMeteors === "true";
+    meteors = meteors.filter(
+      (meteor) =>
+        meteor.is_potentially_hazardous_asteroid === isDangerousMeteors
+    )
+  }
+  if (count && !isNaN(count) && count > 0) {
```
Comment (EugeniBykovski) [inline app/helpers/meteorHelper.js:47]: and here we can write code more clear:
`
  const parsedCount = parseInt(count, 10);
  if (!isNaN(parsedCount) && parsedCount > 0) {
    meteors = meteors.slice(0, parsedCount);
  }
`

**Review by EugeniBykovski — APPROVED** [review]

**Pull request merged**